### ISSUE TITLE
[BUG] Error during alignment: RuntimeError: kaldi::KaldiFatalError

### ISSUE THREAD

**Issue by danablend** [mmcauliffe/kalpy#14]
First of all, thank you for putting in so much work into this wonderful solution. It makes life easier being able to use Python to interface with Kaldi instead of having to string together commands all the time - this is nice!

**Debugging checklist**

[X --> v0.6.0] Have you updated to latest Kalpy version?

**Describe the issue**
During an attempt to make a simple function to align text to audio without any temp files (using GmmAligner.align_utterance), I encountered kaldi::KaldiFatalError during alignment, and couldn't trace the issue further.

**For Reproducing your issue**
All files involved in my testing are in the Drive folder here:
- https://drive.google.com/drive/folders/148-z_-i_7U09QY8HgpUoeAE5_84U0zex?usp=sharing
- The folder `kalpy_testing/gh_raw_files` is a copy paste from this repository's `testing/data` folder.
- The Folder `kalpy_testing/acoustic_model` is the extracted acoustic model files for `English (US) ARPA acoustic model v2.0.0`
- The Folder `kalpy_testing/acoustic_model_mfa` is the extracted acoustic model files for `English MFA acoustic model v2.2.1`
- The `test.py` file is the python script I wrote, which you can run to reproduce the errors.

Extract the entire folder somewhere and make sure the file structure is the same as in the Drive folder.

Environment:
Run `conda install -c conda-forge kalpy` if you don't already have. (I used v0.6.0, since it is the latest version on conda-forge).

Run `python test.py`

This should reproduce the outputs:
```
Running custom forced alignment with no temp files...
Initializing lexicon...
Initializing aligner...
Initializing feature generator...
Performing forced alignment...
Computing features...
Compiling training graph...
Performing alignment...
Traceback (most recent call last):
  File "/app/test.py", line 242, in <module>
    align_one(input_wav, input_align_text, feature_generator, training_graph_compiler, aligner)
  File "/app/test.py", line 155, in align_one
    alignment = aligner.align_utterance(
  File "/env/lib/python3.10/site-packages/kalpy/gmm/align.py", line 72, in align_utterance
    ) = gmm_align_compiled(
RuntimeError: kaldi::KaldiFatalError
```

**Additional context**
I ran all of this in a GPU enabled Docker container on Ubuntu with Python v3.10.

I was ALMOST able to run the test in the repository `test_align_sat_first_pass` using the data in the `kalpy_testing/gh_raw_files/kaldi` folder corresponding to `tests/data/kaldi` in the repo (this is in the python file in the Drive folder I provided). The alignment was generated, but `LexiconCompiler.phones_to_pronunciations` failed to run, dumping the following error:
```
ERROR: SymbolStringToLabels: Symbol "" is not mapped to any integer label in symbol table <unspecified>
ERROR: Failed to compile string `           be          be                                       `, with token_type: symbol
```

I was NOT able to run the test in the repository `test_align`, and got the same `RuntimeError: kaldi::KaldiFatalError` as in my custom function.

I tried with two different dictionary+acoustic combinations:
- English MFA dictionary v2.2.1 + English MFA acoustic model v2.2.1
- English (US) ARPA dictionary v2.0.0 + English (US) ARPA acoustic model v2.0.0

Again, thank you for your work. I have used MFA for the past several months in projects, and am excited to get kalpy to work for even better interfacing with kaldi!

**Comment by danablend** [comment]
Okay, after playing around with it for a little while just now, I managed to set up a script that successfully performs forced alignment with no tmp files:

```
import os
import json

from kalpy.data import Segment
from kalpy.decoder.training_graphs import TrainingGraphCompiler
from kalpy.fstext.lexicon import LexiconCompiler
from kalpy.gmm.align import GmmAligner
from kalpy.feat.mfcc import MfccComputer

from praatio import textgrid as tgio

from _kalpy import feat, transform
from _kalpy.matrix import FloatMatrix
from kalpy.utils import read_kaldi_object

test_dir = "kalpy_testing"

def align_one(
    text: str = "we knew each other in school",
    wav_path: str = "16khz.wav",
    
    lexicon_compiler: LexiconCompiler | None = None,
    training_graph_compiler: TrainingGraphCompiler | None = None,

    lda_mat: FloatMatrix = None,
    feature_generator: MfccComputer | None = None,

    aligner: GmmAligner | None = None,
):
    # Compile training graph
    print("Compiling training graph...")
    training_graph = training_graph_compiler.compile_fst(text)

    # Get features
    print("Getting features...")
    splice_frames = 3
    feats = feature_generator.compute_mfccs_for_export(Segment(wav_path), compress=False)
    feats = feat.splice_frames(feats, splice_frames, splice_frames)
    feats = transform.apply_transform(feats, lda_mat)

    # Perform alignment
    print("Performing alignment...")
    alignment = aligner.align_utterance(training_graph, feats)
    
    # Generate intervals
    print("Generating intervals...")
    intervals = alignment.generate_ctm(aligner.transition_model, lexicon_compiler.phone_table, frame_shift = 0.01)

    # Get actual words
    print("Getting actual words...")
    actual_words = [lexicon_compiler.word_table.find(x) for x in alignment.words]
    text = " ".join(actual_words)
    print(text)

    # Convert phones to pronunciations
    print("Converting phones to pronunciations...")
    ctm = lexicon_compiler.phones_to_pronunciations(text, alignment.words, intervals)
    
    # Get textgrid tiers
    print("Getting textgrid tiers...")
    file_duration = 4.07
    word_tier = tgio.IntervalTier("words", [], minT=0.0, maxT=file_duration)
    for w in ctm.word_intervals:
        word_tier.insertEntry(w.to_tg_interval(file_duration))
    
    return word_tier

if __name__ == '__main__':
    # Kalpy testing to replace MFA for forced alignment
    # https://github.com/mmcauliffe/kalpy/blob/main/tests/test_align.py


    print("Running custom forced alignment with no temp files...")
    align_options = {
        "transition_scale": 1.0,
        "acoustic_scale": 0.1,
        "self_loop_scale": 0.1,
        "beam": 1000,
        "retry_beam": 40,
    }

    # File paths
    input_wav = '16khz.wav'
    input_align_text = 'we knew each other in school'
    
    # acoustic_model_dir = f"{test_dir}/acoustic_model"
    acoustic_model_dir = f"{test_dir}/acoustic_model_mfa"

    acoustic_model_base_path = f"{acoustic_model_dir}/final.mdl"
    acoustic_model_tree_path = f"{acoustic_model_dir}/tree"
    acoustic_model_alignment_path = f"{acoustic_model_dir}/final.alimdl"
    acoustic_lda_mat_path = f"{acoustic_model_dir}/lda.mat"

    # dictionary_path = f"{test_dir}/english_us_arpa.dict"
    dictionary_path = f"{test_dir}/english_mfa.dict"
    lexicon_path = f"{test_dir}/lexicon_fsts.ark"

    model_metadata_path = f"{acoustic_model_dir}/meta.json"

    # Get model metadata
    metadata = json.load(open(model_metadata_path, 'r'))
    phones = metadata['phones']
    features_metadata = metadata['features']

    snip_edges: bool = features_metadata['snip_edges']
    use_energy: bool = features_metadata['use_energy']
    
    # Initialize Lexicon
    print("Initializing lexicon...")
    lexicon_compiler = LexiconCompiler(position_dependent_phones=False, phones=phones)
    lexicon_compiler.load_pronunciations(dictionary_path)

    if not os.path.exists(lexicon_path):
        print("Creating lexicon FSTs")
        lexicon_compiler.fst.write(lexicon_path)

    training_graph_compiler = TrainingGraphCompiler(acoustic_model_base_path, 
                                                    acoustic_model_tree_path,
                                                    lexicon_path,
                                                    lexicon_compiler.word_table)


    # Initialize Aligner
    print("Initializing aligner...")
    aligner = GmmAligner(acoustic_model_alignment_path, **align_options)
    aligner.boost_silence(1.0, lexicon_compiler.silence_symbols)

    # Initialize feature generator
    print("Initializing feature generator...")
    feature_generator = MfccComputer(snip_edges=False)

    # Initialize LDA matrix
    print("Initializing LDA matrix...")
    lda_mat: FloatMatrix = read_kaldi_object(FloatMatrix, acoustic_lda_mat_path)

    # Perform forced alignment.
    print("Aligning one...")
    word_tier = align_one(
        text=input_align_text,
        wav_path=input_wav,
        lexicon_compiler=lexicon_compiler,
        training_graph_compiler=training_graph_compiler,
        aligner=aligner,
        feature_generator=feature_generator,
        lda_mat=lda_mat
    )

    for (interval) in word_tier.entries:
        print(interval)

    print("Forced alignment completed!")
```

Here are some things that caused it to break along the way:
- If lda_mat = None, it didn't work. So I extracted the part that applies lda transformation from the FeatureArchive class and applied it to the extracted features right before performing alignment
- Using ARPA dictionary + ARPA based acoustic model didn't work. I got it to work with English MFA dictionary v2.2.1 + English MFA acoustic model v2.2.1

It takes a while to run `LexiconCompiler.phones_to_pronunciations` after alignment when I want to get the words and their timestamps, but it is very precise. Is there a faster way to obtain the IntervalTier for just the words?

Awesome library, thank you for your work!

**Comment by danablend** [comment]
Update --> I made a mistake which resulted in ARPA not working due to the check `if not os.path.exists(lexicon_path)`. I removed the file check and let it overwrite every time, and it works with ARPA as well now.

ARPA runs significantly faster. About 0.8s to process a 4s audio file, instead of 16s when using the MFA dict.

I'm going to close this - I just had to learn how to use the library properly!

Have a nice day :-)
